### PR TITLE
[JW8-2436] Implement progressive MPEG-TS transmuxing 

### PR DESCRIPTION
**Pull request by johnBartos**
### This PR will...
- Refactor `tsdemuxer.ts` to progressively demux TS segments
    - Splits data into multiples of 188 bytes (size of TS packet)
    - Demuxes "even" data; caches remainder data
        - Appends next data set to remainder on next demux call
    - Demuxes all remaining data on flush
    - Does not drop samples if a keyframe is not present
        - Pushes samples whenever the next AUD is found
        - Video sample drop before a keyframe now occurs in `mp4-remuxer.ts`
- Refactor `mp4-demuxer.ts` to progressively remux elementary streams
    - Requires audio and video samples to be present before generating the an initSegment
    - Requires at least 2 video samples to remux video
        - Duration is computed as the DTS diff between two samples
    - If only 1 audio sample is provided, the duration is computed using the default for its type
        - 1024 for AAC, 1125 for MPEG
    - Remove `audiovideoDeltaDts` logic
        - This code forced a/v to remux at the same DTS, but it unclear why this is necessary
    - Remove empty audio remuxing for entire segments
        - Removed because it is valid to remux video without audio
        - Unclear why this logic was needed - should find test streams for this
- Cleanup & non-functional refactoring

Also includes https://github.com/video-dev/hls.js/pull/2096

### Why is this Pull Request needed?
So that the TSDemuxer and MP4Remuxer can transmux when provided segments as a stream of data. Since `demux` and `remux` is now called multiple times per segment load, they must be able to tolerate not having all required information upfront.

### Are there any points in the code the reviewer needs to double check?
Significant decisions:
- Hls.js will no longer remux empty audio if the segment signals an audio track, but provides no samples. Silent audio is still generated to fill gaps between samples.
- Hls.js no longer aligns the DTS of the first audio/video sample
- Hls.js now requires at least two samples to remux video
- Hls.js uses a default duration when remuxing a single audio sample

I am continuing to verify these decisions to see if any regressions have been introduced.

### Resolves issues:
JW8-2436

### CODE REVIEW

**Comment by johnBartos** [comment]
test this please

File: src/demux/tsdemuxer.ts [hunk 1]
```diff
@@ -34,13 +34,16 @@ const RemuxerTrackIdConfig = {
   text: 4
 };
 
-class TSDemuxer extends NonProgressiveDemuxer {
+class TSDemuxer implements Demuxer {
+  static readonly minProbeByteLength = 1024;
```
Comment (robwalch) [inline src/demux/tsdemuxer.ts]: Shouldn't the min probe length be 188 or a multiple of it? This length requires at least 5 packets to be loaded.

> // a TS fragment should contain at least 3 TS packets, a PAT, a PMT, and one PID, each starting with 0x47

File: src/demux/tsdemuxer.ts [hunk 2]
```diff
@@ -339,20 +372,13 @@ class TSDemuxer extends NonProgressiveDemuxer {
       audioTrack.pesData = audioData;
     }
 
-    if (id3Data && (pes = parsePES(id3Data)) && pes.pts !== undefined) {
-      parseID3PES(pes);
+    if (id3Data && (pes = this._parsePES(id3Data)) && pes.pts !== undefined) {
+      this._parseID3PES(pes);
```
Comment (robwalch) [inline src/demux/tsdemuxer.ts]: What's the underscore for?

How about declaring this as a private class method?
`private parseID3PES(pes: PacketizedElementaryStream) {`

Is the `const parseID3PES = ...bind(this)` still used?

File: src/demux/tsdemuxer.ts [hunk 3]
```diff
@@ -609,7 +620,7 @@ class TSDemuxer extends NonProgressiveDemuxer {
       push,
       spsfound = false,
       i,
-      pushAccesUnit = this.pushAccesUnit.bind(this),
+      pushAccessUnit = this.pushAccessUnit.bind(this),
```
Comment (robwalch) [inline src/demux/tsdemuxer.ts]: What are these for? I'm not seeing the context change that would require a bind. Am I missing something?

Comment (johnBartos) [reply src/demux/tsdemuxer.ts]: Legacy code, I suppose to avoid using `this` a lot

Comment (robwalch) [reply src/demux/tsdemuxer.ts]: I guess that's one way to get the code to minify, but creating new runtime functions inside code that runs over and over isn't the most efficient thing.

**Pull request merged**